rl4co/envs/routing/tsp.py

```diff
@@ -161,6 +174,50 @@ def check_solution_validity(td: TensorDict, actions: torch.Tensor):
             == actions.data.sort(1)[0]
         ).all(), "Invalid tour"
 
+    @staticmethod
+    def improve_solution(td: TensorDict, actions: torch.Tensor, **kwargs) -> torch.Tensor:
+        """Improve the solution using local search"""
+        assert pyvrp is not None, (
+            "`pyvrp` not found. Please install pyvrp using instructions from "
+            "https://github.com/PyVRP/PyVRP or just use pip install pyvrp."
+        )
+        
+        rng = RandomNumberGenerator(kwargs.get("seed") or 0)
+        
+        action_list = actions.cpu().detach().numpy().tolist()
+        scaled_locs = scale_and_truncate_to_decimals(td["locs"].cpu().detach().numpy(), 7)
+        scaled_distmat = scale_and_truncate_to_decimals(
+            get_distance_matrix(td["locs"]).cpu().detach().numpy(), 7
+        )
+        
+        improved_action_list = []
+        for _locs, _distmat, _acts in zip(scaled_locs, scaled_distmat, action_list):
```

Comment:
If we have a batch, would it be better to call `concurrent.futures`?

Reply:
As pyvrp doesn't support TwoOpt in recent releases, I will replace it with the [implementation of DeepACO](https://github.com/henry-yeh/DeepACO/blob/main/tsp_nls/two_opt.py), which uses `concurrent.futures`.